File: src/nack.py
import sys
import subprocess
import re
import json

def parse_nmap_output(stdout):
    result = {}
    lines = stdout.split('\n')
    current_ip_address = ""

    for line in lines:
        if 'Nmap scan report for' in line:
            current_ip_address = line.split(' ')[4]
            result[current_ip_address] = {}
        elif ('open' in line or 'filtered' in line or 'closed' in line) and '/' in line:
            split_line = line.split()
            port_protocol = split_line[0].split('/')
            port = int(port_protocol[0])
            protocol = port_protocol[1]
            state = split_line[1]
            service = split_line[2]

            result[current_ip_address][port] = {
                'state': state,
                'protocol': protocol,
                'service': service,
            }

    return result

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Please provide an IP address, range of IPs, or a domain as a command-line argument.")
        sys.exit(1)

    target = sys.argv[1]
    command = f"nmap -p- {target}"

    process = subprocess.Popen(command.split(), stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    stdout, stderr = process.communicate()

    if process.returncode != 0:
        print(f"exec error: {stderr.decode()}")
        sys.exit(1)

    open_ports = parse_nmap_output(stdout.decode())
    print(json.dumps(open_ports))

File: src/nack.ts
const { exec } = require("child_process");

// Retrieve target from the command line arguments
const target = process.argv[2];

if (!target) {
  console.error("Please provide an IP address, range of IPs, or a domain as a command-line argument.");
  process.exit(1);
}

// Run nmap command to scan for open ports
exec(`nmap -p- ${target}`, (error, stdout, stderr) => {
  if (error) {
    console.error(`exec error: ${error}`);
    return;
  }

  const openPorts = parseNmapOutput(stdout);
  console.log(JSON.stringify(openPorts));
});

interface PortInfo {
  state: string;
  protocol: string;
  service: string;
}

function parseNmapOutput(stdout: string): Record<string, Record<number, PortInfo>> {
  const result: Record<string, Record<number, PortInfo>> = {};
  const lines = stdout.split('\n');
  let currentIpAddress = "";

  lines.forEach(line => {
    if (line.includes('Nmap scan report for')) {
      currentIpAddress = line.split(' ')[4];
      result[currentIpAddress] = {};
    } else if (line.includes('open') || line.includes('filtered') || line.includes('closed')) {
      const splitLine = line.split(/\s+/);
      const portProtocol = splitLine[0].split('/');
      const port = parseInt(portProtocol[0], 10);
      const protocol = portProtocol[1];
      const state = splitLine[1];
      const service = splitLine[2];

      result[currentIpAddress][port] = {
        state: state,
        protocol: protocol,
        service: service,
      };
    }
  });

  return result;
}


module.exports = { parseNmapOutput };